open-apparel-registry/open-apparel-registry
# Add facility indexing

## Overview

Adds an indexed table of facilities, incorporating fields which are
utilized by the facilities search. Data such as facility contributors,
which require several relationship traversals to find, is denormalized
into array fields. A GinIndex is used to index the array fields.

Adds a management command to index facilities. To index all facilities,
pass no arguments to the management command. To index one or more
facilities, pass the ids of the facilities to be indexed.

Adds post_save signals to Facility, Source, FacilityMatch, and Contributor to 
index / re-index facilities after related entities have been updated or created. 

Updates the facilities search to utilize the new search index. In a test on staging, 
the new indexed search took an average of ~0.298 seconds. In comparison, 
the previous search took an average of ~0.617 seconds. 

Connects #1433 

## Demo

Queries run on Amazon Athena to analyze the time taken to search:

BEFORE
`SELECT AVG(target_processing_time) as Average, COUNT(target_processing_time) as Count FROM stg_alb_logs WHERE time > '2021-08-17' AND time < '2021-08-18T14:00:24.062714Z' AND request_url LIKE '%tile%';`
Average: 0.6175616475722862
Count: 3666

AFTER
`SELECT AVG(target_processing_time) as Average, COUNT(target_processing_time) as Count FROM stg_alb_logs WHERE time > '2021-08-18T14:30:24.062714Z' AND request_url LIKE '%tile%';`
Average: 0.2980567260940032
Count: 617

## Notes

- We use ArrayFields to store the related data for the facilities. In order to index these, we use a GinIndex. More info: [Django Docs](https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/fields/), [PostgreSQL Docs](https://www.postgresql.org/docs/current/gin-intro.html)
- In order to keep the search changes localized, the search creates a list of facility ids and then returns all of the Facilities with those ids. Testing showed that altering the search to return and the tile search to use the FacilityIndexes directly (avoiding the Facility mapping) was slower than using the Facility mapping. 
- Running the queries produced by the search against `explain analyze` in the dbshell showed that the FacilityIndex search wasn't making use of the GinIndex because on the current dataset it would be slower than use a sequential scan. However, when we turned off sequential scanning functionality in the database, the GinIndex was used successfully (but was in fact slightly slower). It seems likely that with a larger volume of FacilityIndexes the GinIndex will become the faster option and be used in the search as expected.
- The indexes are updating based on post_save signals. Using alternative update methods such as bulk update might in some cases not initiate the re-indexing of facilities.

## Testing Instructions

* Run `./scripts/manage migrate`. The migration should run correctly.
* Run `./scripts/manage index_facilities`.
* Run the search with various combinations of filters. Confirm that the search behaves as expected, and that the tiles and facilities returned are correct.  
* Update contributors, facility details, sources, and facility matches to various states using the Django admin. Confirm that the search reflects the changes to the data. 

## Checklist
